File: 2024/February/Special paths.cpp
#include <iostream>
#include <cstring>

const int maxn = 1e5 + 14;

struct Dsu {
    int par[maxn];

    Dsu() {
        std::memset(par, -1, sizeof par);
    }

    int root(int v) {
        return par[v] < 0 ? v : par[v] = root(par[v]);
    }

    bool fri(int v, int u) {
        return root(v) == root(u);
    }

    bool merge(int v, int u) {
        if ((v = root(v)) == (u = root(u)))
            return 0;
        par[u] += par[v];
        par[v] = u;
        return 1;
    }
} dsu;

int n, m, v[maxn], u[maxn], a[maxn];

int main() {
    std::ios::sync_with_stdio(0);
    std::cin.tie(0);

    std::cin >> n >> m;

    for (int i = 0; i < m; ++i) {
        std::cin >> v[i] >> u[i];
        v[i]--, u[i]--;
    }

    for (int i = 0; i < n; ++i) {
        std::cin >> a[i];
    }

    int s, e;
    std::cin >> s >> e;
    s--, e--;

    int lo = -1, hi = 1e6;

    while (hi - lo > 1) {
        dsu = Dsu();
        int mid = (lo + hi) / 2;

        for (int i = 0; i < m; ++i) {
            if (std::abs(a[v[i]] - a[u[i]]) <= mid)
                dsu.merge(v[i], u[i]);
        }

        (dsu.fri(s, e) ? hi : lo) = mid;
    }

    std::cout << hi << '\n';

    return 0;
}

File: 2024/February/Special paths.py
class Dsu:
    def __init__(self, n):
        self.par = [-1] * n
    
    def root(self, v):
        return v if self.par[v] < 0 else self.root(self.par[v])
    
    def is_same_set(self, v, u):
        return self.root(v) == self.root(u)
    
    def merge(self, v, u):
        v, u = self.root(v), self.root(u)
        if v == u:
            return False
        self.par[u] += self.par[v]
        self.par[v] = u
        return True

def main():
    n, m = map(int, input().split())
    dsu = Dsu(n)
    
    v, u = [], []
    for i in range(m):
        vi, ui = map(int, input().split())
        v.append(vi - 1)
        u.append(ui - 1)

    a = list(map(int, input().split()))
    
    s, e = map(int, input().split())
    s, e = s - 1, e - 1
    
    lo, hi = -1, 10**6
    while hi - lo > 1:
        dsu = Dsu(n)
        mid = (lo + hi) // 2
        for i in range(m):
            if abs(a[v[i]] - a[u[i]]) <= mid:
                dsu.merge(v[i], u[i])
        
        if dsu.is_same_set(s, e):
            hi = mid
        else:
            lo = mid
    
    print(hi)

if __name__ == "__main__":
    main()
